Switch from Autotools to CMake+Ninja

klaasvanaarsen opened the pull request:
I propose switching from Autotools to CMake + Ninja as the primary build system. The foundation was already well laid by previous contributors - kudos to them!

## Benefits

- 2x faster builds (10s vs 27s on GitHub Actions)
- Smaller binary size
- Less maintenance (by eventually removing Autotools)
- Modern standard used by most C/C++ projects

## What This PR Does
- Adds CMake + Ninja builds to all Windows CI steps:
- Linux debug cross-build
- Msys2 build
- Release build

The results demonstrate consistent performance improvements across all platforms.

-----

### Windows (linux debug cross-build for Windows)

<img width="348" height="254" alt="Screenshot 2025-11-02 233546" src="https://github.com/user-attachments/assets/2c59d35b-2e3c-48ca-b4bf-fddda56bce03" />

### create-github-release (linux release cross-build for Windows)

<img width="433" height="425" alt="Screenshot 2025-11-02 233741" src="https://github.com/user-attachments/assets/eda043cd-7fa6-4aa7-9665-ccccf2f6f442" />

klaasvanaarsen commented:
**Summary:** Autotools effectively uses `-O0` in `release.yaml`, which explains the difference in executable size.

**Details:**
1. CMake defaults to `-O3` unless overridden. In release.yaml we explicitly set `-O2` to match autotools’ intended optimization. However, the autotools compile command shows `-O0`, which likely explains the size difference.
2. The CMake build produces a console (CUI) executable, while autotools produces a GUI executable. This should not affect Windows compatibility. We can configure CMake to produce a GUI executable if exact parity is desired.
3. There are minor differences in imported functions (e.g., `FindClose` vs `_findclose`) likely due to the GUI/CUI distinction.
4. Some internal symbols differ due to the different compiler flags and configuration; see the compile commands below. We can align these if strict parity is required.
5. The list of external DLLs is identical, though the order differs.
6. Both builds target Windows version 4.0 (Windows 95/NT 4.0), so there is no regression in OS coverage.

## Compile and link commands

### autotools
```
i686-w64-mingw32-gcc -O2 -DNDEBUG -W -Wall -Wextra -Wno-unused-parameter -pedantic -Wstrict-prototypes -Wmissing-prototypes -Wwrite-strings -Wnested-externs -Wshadow -static -DTEST_OVERRIDE_PATHS  -DHAVE_VERSION_H -I. -fPIC -std=c99 -O0 -DHAVE_CONFIG_H -I. -DWINDOWS -Iwin/include   -MD -MF ${out%.o}.dep -c -o cave.o cave.c
	
i686-w64-mingw32-gcc -o angband.exe win/angband.res main-win.o win/readdib.o win/readpng.o win/scrnshot.o win/win-layout.o angband.a -Lwin/lib   -mwindows -lwinmm -lzlib -llibpng -lmsimg32
```

### cmake
```
i686-w64-mingw32-gcc -DALLOW_BORG -DBUILD_ID=4.2.5-492-gb570f6223-dirty -DSOUND -DUSE_PRIVATE_PATHS -DUSE_WIN -DWINDOWS -D_CRT_SECURE_NO_WARNINGS @CMakeFiles/OurCoreLib.dir/includes_C.rsp -O2 -DNDEBUG -std=gnu99 -MD -MT CMakeFiles/OurCoreLib.dir/src/cave.c.obj -MF CMakeFiles/OurCoreLib.dir/src/cave.c.obj.d -o CMakeFiles/OurCoreLib.dir/src/cave.c.obj -c /mnt/c/GAMES/Angband/angband/src/cave.c

i686-w64-mingw32-gcc -O2 -DNDEBUG -mwindows -Wl,--whole-archive CMakeFiles/OurExecutable.dir/objects.a -Wl,--no-whole-archive -o Angband.exe -Wl,--out-implib,libAngband.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles/OurExecutable.dir/linklibs.rsp
```

## Ninja documentation

This PR also adds a section in `docs/hacking/compiling.rst` describing how to use Ninja with CMake for faster builds now.

klaasvanaarsen commented:
I've extended all YAML files with a CMake/Ninja step where it wasn’t already present (and where CMakeLists.txt supported it), and I’ve updated existing invocations to use Ninja.

From what I can tell, all current development work is already happening with CMake — the main build steps use it. The only remaining parts using autotools are the release process, MacOS, specialized builds, and build steps for old systems.

That creates an undesirable situation: the release build is not the one being actively developed. In particular, the compilation flags have diverged, meaning that what contributors develop and test may not be what ends up in the release.

So I think we should switch to CMake as the primary build system — at least to ensure that contributions are built and tested the same way they’ll appear in the release.

I’ve updated `docs/hacking/compiling.rst` to describe using Ninja (and multiple cores), which also applies to make. The MSYS2 section now uses the toolchain included in the repo.

If we decide to make CMake the primary system, I’ll:
* Update `release.yaml` to use CMake instead of autotools
* Adjust `Compiling.rst` to document CMake first and autotools second

Otherwise, I’ll revert the changes to `release.yaml` (except for the multicores update, which also benefits make).

After that, this PR should be ready to merge.

What do you think, @backwardsEric? Could you take a look?

backwardsEric commented:
The use of CMake versus autotools in the existing workflows files is largely the result of spur of the moment choices rather than a coordinated design.  The original author of linux.yaml chose to use CMake and most of what happens there still uses CMake.  windows.yaml and the Windows portion of release.yaml were originally written when CMakeLists.txt did not support building the Windows front end so they use autotools.

Judging from posts in the forums and submitted issues and pull requests, people compiling Angband use autotools more than CMake.  Changing how the different build systems are promoted in docs/hacking/compiling.rst might help with that.  In my personal use, most builds use src/Makefile.osx.  When I use autotools or CMake and am not specifically working on something specific to one of those build systems, I have a slight preference for using autotools.  Building with Angband with CMake is still rougher than using autotools:  witness the changes needed in this pull request to smooth out problems using CMake.

As you note, not exercising the release compiler flags except in release.yaml is a problem with the current set of workflows.  Adding new parts to windows.yaml and mac.yaml which use those flags would be enough, in my opinion, to catch problems when a pull request is submitted.  That's outside of the scope of this pull request so I would not modify what you have here for that.

As for what I would like to see changed in this pull request:

- The change at lines 802-807 of CMakeLists.txt renames Angband.exe to angband.exe when copying over the executable.  I think it makes more sense to make a change at line 3 of CMakeLists.txt so "SET(OUR_EXECUTABLE_NAME ${PROJECT_NAME})" is instead "STRING(TOLOWER ${PROJECT_NAME} OUR_EXECUTABLE_NAME)":  that would have the results of a build with CMake match the names used in Makefile.src and avoid adding another place in CMakeLists.txt where a variant would have to replace "angband" or "Angband" with whatever name was appropriate for the variant.  With the pull request as is or using angband as the executable name on all platforms, changes will have to be made in docs/hacking/compiling.rst to be consistent with the changes to the name.  If angband is used on all platforms, need to change "To run the result, change directories to build (or whatever you named it to) and run ./Angband ." around docs/hacking/compiling.rst:178 to replace "./Angband" with "./angband" and change "That will leave an Angband.exe and the needed .dll files in the directory where cmake was run.  That executable can be run with wine:  wine Angband.exe" around docs/hacking/compiling.rst:350 to replace "Angband.exe" with "angband.exe".
- As I said in my earlier comment, I would either not add the cmake_build portion to release.yaml or use it replace the portion that uses autotools to cross-compile the Windows executable for the release.  If you replace the autotools part, check that the CMake build does not make use of -march (or other compiler flags that affect what instruction set is used) that would limit binary compatibility more than what the autotools build does.

klaasvanaarsen commented:
- I have restored the section about OpenBSD.
- I have restored the behavior that the windows frontend message is given only once.
- I found out more about the mysterious compilation failure on MSYS2. It turns out that it has its own zlib1.dll, which is different from the bundled one. If the bundled zlib1.dll is present in the build directory, compilation of any .c file fails with exit status 1 and no diagnostics, even though the compilation does not refer to the dll. The workaround to delete the DLLs works on CI, but it is fragile in development. I'm looking for a more robust solution.

backwardsEric commented on .github/workflows/linux.yaml:
```diff
@@ -90,7 +90,7 @@ jobs:
         run: |
           sudo apt-get update
-          sudo apt-get install gcc autoconf automake make libsdl2-dev libsdl2-image-dev libsdl2-ttf-dev libsdl2-mixer-dev
+          sudo apt-get install gcc autoconf automake make libsdl2-dev libsdl2-image-dev libsdl2-ttf-dev libsdl2-mixer-dev ninja-build
```
I don't think it's necessary to have the SDL2 version built with both autotools and CMake here.  Remove the dependencies that are only needed for the autotools build and remove the "Build" step at lines 98-102 that uses autotools.

klaasvanaarsen commented:
- I have moved the game files to the `game/` subdirectory and updated `Compiling.rst`, since that is necessary to get MSYS2 to work cleanly as long as we want to support bundled dll's on MSYS2. Note that the tests also land in that game/ directory now as they depend on the exe and the lib directory.
- I have added `cmake` in the MSYS2 section of `Compiling.rst` before the explanation with `make`.
- As yet I have not added SUPPORT_WINDOWS_Frontend to MSYS2, since that currently depends on MINGW32 while the section is specifically about MINGW64.

backwardsEric commented on docs/hacking/compiling.rst:
```diff
@@ -177,7 +177,7 @@ one or more of the other graphical front ends are selected. The graphical front
 ends are: GCU, SDL, SDL2 and X11.  All of the following generate a
 self-contained directory, build, that you can move elsewhere or rename.  To
```
Since the executable is now in build/game.  This should be "self-contained directory, build/game, that you can move elsewhere or rename.  To run the result, change directories to build/game or whatever you renamed it to) and run ./angband ."

klaasvanaarsen replied:
Updated.
I have also added Windows as a graphical front end here.

klaasvanaarsen commented on .github/workflows/linux.yaml:
```diff
@@ -13,18 +13,18 @@ jobs:
         run: |
           sudo apt-get update
-          sudo apt-get install gcc make cmake libx11-dev
+          sudo apt-get install gcc make cmake libx11-dev ninja-build
```
No, make is not needed anymore if it is not referred to in the step. So far I left it to remain flexible and not worry whether make is there or not - it will "just work". Since it is a tiny package it doesn't hurt to have it around. Then again, I guess there is something to be said to make it as lean and clean as possible and protect against an accidental use of make. Is that what you prefer?

backwardsEric replied:
I prefer only including the direct dependencies and nothing more.  If the workflow is used as an example of what the prerequisites are for building Angband on <x>, then all someone sees is the minimum set of things that will be needed to do the build.

klaasvanaarsen commented:
- Implemented SUPPORT_BUNDLED_PNG and added cmake build for it to MSYS2/MINGW32
- Added WINDOWS_Frontend cmake build to MSYS2/MINGW64 that uses the system PNG.
- Added Windows as a target in the Cmake section in Compiling.rst.
- Added cmake command for SUPPORT_WINDOWS_Frontend to the MSYS2/MINGW64 section in Compiling.rst.

NickMcConnell commented:
What's the status of this one?

klaasvanaarsen commented on src/cmake/modules/PNG_Detection.cmake:
```diff
@@ -0,0 +1,85 @@
+        INTERFACE_INCLUDE_DIRECTORIES "${PNG_INCLUDE_DIRS}"
+    )
+
+    list(LENGTH PNG_LIBRARIES COUNT)
```
Currently the code does indeed assume that the first library is libpng and the second is zlib. Because those are assumptions, there’s an explicit sanity check to verify that this ordering is correct before proceeding.

It’s certainly possible to iterate over all libraries returned by find_package(PNG), identify which one is the actual PNG library, which one is zlib, and also handle edge cases such as additional dependencies or libpng builds that do not require zlib at all. That would be more robust, but it also requires a fair amount of extra logic that we don’t currently need. Given the existing checks and the limited scope of what we’re trying to solve here, I’ve kept the implementation simple for now.

backwardsEric replied:
I'm still trying to understand if we need or want to copy the non-bundled versions of libpng and zlib and if they need to be represented by imported library targets.  For the MSYS2 case, does the compiled Windows front end run if the MSYS2 packaged versions of the DLLs are not copied over?  For the case where someone is linking against their own copies of those libraries, they presumably know what they are doing and know that they normally need to copy over whatever DLLs are necessary.  So the copy step in that case is convenient, but not necessary.

If we can dispose of the copy step for non-bundled DLLs and don't need imported library targets for them, we wouldn't need DeduceDLL.cmake and PNG_Detection.cmake would be simpler and could be folded not WINDOWS_Frontend.cmake.

klaasvanaarsen replied:
We can run angband within the MSYS2 shell where we built it without copying the DLLs.
But if we open the angband executable with for instance the Explorer, it can't find the DLLs if they are not in the same directory.

klaasvanaarsen replied:
I understand and agree that this involves a lot of CMake code to figure out the DLLs, and that it relies on some assumptions.

Ideally, `find_package(PNG)` would not only provide the import libraries (IMPORTED_IMPLIB) but also the runtime DLLs (IMPORTED_LOCATION). Unfortunately, it doesn’t, and there seems to be no simpler way to achieve this.

This appears to be by design: CMake’s `find_package` is intended to locate build-time dependencies and deliberately keeps redistribution and packaging out of scope, which makes sense given licensing considerations. That said, in our case we do need the correct libpng DLLs to package, and licensing is not an issue.

backwardsEric replied:
When possible, I would prefer that the user chooses whether to statically or dynamically link external libraries that are not bundled:  that way the user can better match their requirements.

When I checked the results of your msys2.yaml with ntldd, it indicated that the CMake build for SDL2 dynamically linked SDL2 (which is what I expect, since statically linking it and its dependencies, like Makefile.msys2.sdl2 does, is a nightmare).  At least for now, that workflow result is here, https://github.com/backwardsEric/angband/actions/runs/19571659815 .  So the executable from that build should not work outside of the MSYS2 environment without extra work:  the same issue that you are trying to solve for the build with CMake on MSYS2 of the Windows front end.

Since we don't have a general solution for that problem, I would say that users of MSYS2 are on their own if they want to run the result of a build outside of the MSYS2 environment or want to make the result of cmake --install useful.  The build of the Windows front end with the bundled DLLs on MSYS2 avoids those complications if the user is satisfied with the Windows front end and a 32-bit x86 executable.

Putting the executable, DLLs, and lib in the game subdirectory of the build directory seemed like an acceptable tradeoff to me.  It is a breaking change for someone expecting the executable in the build directory but it avoids an issue that I could see happening under other situations on Windows:  CMake sets the current working directory for the compiler to the build directory, the compiler depends on zlib1.dll, and with an incompatible (probably 32-bit versus 64-bit) zlib1.dll present in the working directory, the compiler fails to start, possibly with no useful diagnostic messages.

So, I'd prefer a solution that:

- Does not force the user to statically link an external library.
- When not using the bundled DLLs, just uses find_package(PNG) and does not try to go farther than that to copy over or install any DLLs.
- Allows SUPPORT_BUNDLED_PNG=ON to work on MSYS2.

klaasvanaarsen replied:
Done.

On the topic of `cmake --install`, we now have:

```
MSYS2 MINGW32 $ cmake -G Ninja -DSUPPORT_BUNDLED_PNG=ON ..
MSYS2 MINGW32 $ ninja install
[216/217] Install the project...-- Install configuration: ""
-- Installing: C:/Program Files (x86)/Angband/./angband.exe
-- Up-to-date: C:/Program Files (x86)/Angband/./libpng12.dll
-- Up-to-date: C:/Program Files (x86)/Angband/./zlib1.dll
-- Up-to-date: C:/Program Files (x86)/Angband/lib
-- Up-to-date: C:/Program Files (x86)/Angband/lib/.deps
-- Up-to-date: C:/Program Files (x86)/Angband/lib/customize
-- Up-to-date: C:/Program Files (x86)/Angband/lib/customize/.deps
-- Up-to-date: C:/Program Files (x86)/Angband/lib/customize/font-gcu.prf

MSYS2 MINGW32 $ cmake -G Ninja -DSUPPORT_BUNDLED_PNG=OFF .. && ninja
MSYS2 MINGW32 $ ninja install
[1/2] Install the project...-- Install configuration: ""
-- Installing: C:/Program Files (x86)/Angband/./angband.exe
-- Up-to-date: C:/Program Files (x86)/Angband/lib
-- Up-to-date: C:/Program Files (x86)/Angband/lib/.deps
-- Up-to-date: C:/Program Files (x86)/Angband/lib/customize
-- Up-to-date: C:/Program Files (x86)/Angband/lib/customize/.deps
-- Up-to-date: C:/Program Files (x86)/Angband/lib/customize/font-gcu.prf
```

So `cmake --install` works as before if we use the bundled PNG.
It currently still installs if we use the system PNG, but it is effectively broken since the DLLs are missing.
As far as I can tell, it would require some cmake logic based on assumptions to eliminate the install in the latter case, so I guess it's best to just live with it.

On a side note, the `.deps` files are coming along when they were supposed to be filtered out. It appears that filtering logic is broken. I'll leave that out of scope for this PR though.

backwardsEric commented on docs/hacking/compiling.rst:
```diff
@@ -206,6 +207,12 @@ To build Angband with the GCU front end::
     cmake -DSUPPORT_GCU_FRONTEND=ON ..
     make
 
+To build Angband with the Windows front end::
```
As above, this is for Linux/other Unix so need cross-compiling for Windows.  So I'd just refer to the cross-compiling for Windows part here.

klaasvanaarsen replied:
Reverted

backwardsEric commented:
I would leave the built documentation where it is.  It appears where it does in the zip file for Windows because of scripts/pkg_win.
